cbmc/proofs/mqtt_state.c

```diff
@@ -0,0 +1,556 @@
+ * IotMqttNetworkInfo
+ ****************************************************************/
+
+IotMqttNetworkInfo_t *allocate_IotMqttNetworkInfo(IotMqttNetworkInfo_t *pInfo)
+{
+  if ( pInfo == NULL ) pInfo = malloc_can_fail( sizeof( *pInfo ) );
+  if ( pInfo == NULL ) return NULL;
+
+  if (pInfo->createNetworkConnection) {
+    // allocate setup member of union
+    pInfo->u.setup.pNetworkServerInfo = allocate_opaque_type();
+    pInfo->u.setup.pNetworkCredentialInfo = allocate_opaque_type();
+  } else {
+    // allocate network member of union
+    pInfo->u.pNetworkConnection = allocate_IotNetworkConnection();
+  }
+  pInfo->pNetworkInterface = allocate_IotNetworkInterface();
+  return pInfo;
+}
+
+bool valid_IotMqttNetworkInfo(IotMqttNetworkInfo_t *pInfo)
+{
+  if ( pInfo == NULL ) return 0;
+  return 1;
+}
+
+/****************************************************************
+ * IotMqttConnectInfo
+ ****************************************************************/
+
+IotMqttConnectInfo_t *allocate_IotMqttConnectInfo(IotMqttConnectInfo_t *pInfo)
+{
+  if ( pInfo == NULL ) pInfo = malloc_can_fail( sizeof( *pInfo ) );
+  if ( pInfo == NULL ) return NULL;
+
+  pInfo->pPreviousSubscriptions =
+    allocate_IotMqttSubscriptionArray(NULL,
+				     pInfo->previousSubscriptionCount);
+  pInfo->pWillInfo = allocate_IotMqttPublishInfo(NULL);
+  pInfo->pClientIdentifier = malloc_can_fail(pInfo->clientIdentifierLength);
+  pInfo->pUserName = malloc_can_fail(pInfo->userNameLength);
+  pInfo->pPassword = malloc_can_fail(pInfo->passwordLength);
+}
+
+bool valid_IotMqttConnectInfo(IotMqttConnectInfo_t *pInfo)
+{
+  return
+    pInfo &&
+
+    VALID_STRING(pInfo->pClientIdentifier, pInfo->clientIdentifierLength) &&
+    VALID_CBMC_SIZE( pInfo->clientIdentifierLength ) &&
+
+    VALID_STRING(pInfo->pUserName, pInfo->userNameLength) &&
+    VALID_CBMC_SIZE( pInfo->userNameLength ) &&
+
+    VALID_STRING(pInfo->pPassword, pInfo->passwordLength) &&
+    VALID_CBMC_SIZE( pInfo->passwordLength ) &&
+
+#ifdef SUBSCRIPTION_COUNT_MAX
+    pInfo->previousSubscriptionCount < SUBSCRIPTION_COUNT_MAX &&
+#endif
+    IFF( pInfo->pPreviousSubscriptions == NULL,
+	 pInfo->previousSubscriptionCount == 0 ) &&
+    valid_IotMqttSubscriptionArray(pInfo->pPreviousSubscriptions,
+				   pInfo->previousSubscriptionCount);
+}
+
+/****************************************************************
+ * IotMqttSubscription
+ ****************************************************************/
+
+IotMqttSubscription_t *allocate_IotMqttSubscription(IotMqttSubscription_t *pSub)
+{
+  if ( pSub == NULL ) pSub = malloc_can_fail( sizeof( *pSub ) );
+  if ( pSub == NULL ) return NULL;
+
+  pSub->pTopicFilter = malloc_can_fail( pSub->topicFilterLength );
+  return pSub;
+}
+
+bool valid_IotMqttSubscription(IotMqttSubscription_t *pSub)
+{
+  return
+    pSub &&
+
+    VALID_STRING( pSub->pTopicFilter,  pSub->topicFilterLength ) &&
+    VALID_CBMC_SIZE( pSub->topicFilterLength )
```

Comment:
I wonder if the combination of these two could be a macro?

Reply:
I don't think so.  The valid_string should be true of all strings, not just the ones we allocate, and I think it is more clear as is.